Add target_hash to StochasticLosses

Jean1995 opened the pull request:
This adds information about the component that we interacted with to the Secondaries object and the StochasticLoss object.
If we interacted with a component, we can use the `GetComponent` method to directly access the Component. This can then be used to sample secondary particles (solves issue #221).

In a future release, we might want to replace the hash with a shared_ptr to the original Component since this is more transparent for the user. However, we would need to restructure some parts of  the `Interaction` (and `Interaction::Loss`) and also `CrossSection` classes since all of these classes work with the hashes instead of pointer.

sudojan commented on src/PROPOSAL/detail/PROPOSAL/particle/Particle.cxx:
```diff
@@ -125,10 +126,20 @@ double ParticleState::GetMomentum() const
+                               propagated_distance(propagated_distance),
+                               target_hash(target_hash) {}
+
+Component StochasticLoss::GetComponent() const {
```
How to get the medium?

Jean1995 replied:
I added a `medium_map` to the `Medium` class to get the corresponding Medium to a hash.
The `GetComponentForHash()` method has also been moved to be a static function of the Component class, there is also a `GetMediumForHash()` method now in the Medium class.

I think for the future, we should use pointers instead of these global maps.

sudojan replied:
> I think for the future, we should use pointers instead of these global maps.

well, then create an issue

sudojan reviewed (changes requested):
An example script or Unit Test is missing.

Jean1995 commented:
> An example script or Unit Test is missing.

I added an example on how to extract the StochasticLosses and the information about the target hash/medium to the Propagator jupyter notebook example.

I also added UnitTest for the MediumMap and the ComponentMap to check whether all objects have been correctly registered.

sudojan commented:
> The ID of the Interaction Type is 1000000004. The hash of the target that we interacted with is 9764564051507188568.

In the example notebook of the propagator, this output could be a little bit more descriptive. How to get component/medium name out of the hash is already shown afterwards, but this is not done for the interaction type. Although this might already be shown in another example notebook, it would be practical to have this also here.

sudojan commented on src/PROPOSAL/PROPOSAL/Secondaries.h:
```diff
@@ -51,11 +51,13 @@ class Secondaries {
-    void push_back(const ParticleState& point, const InteractionType& type);
+    void clear() { track_.clear(); types_.clear(); target_hashes_.clear(); };
+    void push_back(const ParticleState& point, const InteractionType& type,
+                   const size_t& target_hash = 0);
```
Why does the hash get a default value?

Jean1995 replied:
Mainly for backward compatibility for people who might have used the `push_back` function without specifying a target_hash

Jean1995 commented:
> > The ID of the Interaction Type is 1000000004. The hash of the target that we interacted with is 9764564051507188568.
> 
> In the example notebook of the propagator, this output could be a little bit more descriptive. How to get component/medium name out of the hash is already shown afterwards, but this is not done for the interaction type. Although this might already be shown in another example notebook, it would be practical to have this also here.

Done. This was a very good idea, since it was not even clear to me how to extract this information in python 🙈

sudojan commented on examples/Propagator.ipynb:
```diff
@@ -281,6 +281,123 @@
+    "We can see that our `StochasticLoss` objects carries the information about the target that we interacted with in the form of a hash.\n",
+    "For Ionization interactions, this corresponds to the hash of a Medium object.\n",
+    "For all other interactions, this corresponds to the hash of a Component object (i.e. an atom).\n",
+    "We can get the Medium/Component object corresponding to these hashes using the methods `pp.medium.medium_map()` and `pp.component.component_map()`:"
```
Due to the recent changes, this comment should refer to the methods `pp.medium.get_medium_for_hash()` and `pp.component.get_component_for_hash()`
